### PR TITLE
_checkLinesLength (GOKbIT-247)

### PR DESCRIPTION
**Pull request by philboeselager**
Just a draft for now.
`_checkLinesLength()` is not executed, yet.
Can you tell why, @hornmo ?

### CODE REVIEW

**Review by hornmo — COMMENTED** [review]
At the point you are calling the method at, `this.readerForImport` is not yet set. Methods using the result of the read action have to be triggered via `FileReader.onload`. In this case there is a separate read action for the charset detections, since we need to check the binary string there. 

So everything you do you would have to do inside of either `onload` method, which probably means extending `_checkCharset`, since we want to do this before we start the second read process.

**Comment by philboeselager** [comment]
@hornmo : your last comment was fixed in [5f73dc6](https://github.com/openlibraryenvironment/gokb-ui/pull/71/commits/5f73dc69077e00de8ec8842abc010835460e235e). 
In [265028e](https://github.com/openlibraryenvironment/gokb-ui/pull/71/commits/265028ee84ab4b14df09f0b8142708a295f75214), I went for allowing multiple error messages, since the most recent one had always been overwriting the previous one. But the changes break some style (line length does not affect the div width anymore) and the detail error table below does not appear anymore. I can't see why. Can you?

**Comment by hornmo** [comment]
I think it would be best to just replace error with errors completely and have something like
`      <div
        v-for="er in errors"
        class="ma-2"
      >
        {{ er }}
      </div>`
instead.

Also in my test I noticed that a KBART file I used had another tab at the end of each line (except the first one) and therefore all rows threw an error (with 27 instead of 26 columns counted). Not sure if we want to ignore this and instead only look at fewer tabs.

I found one other issue not connected to this in `ensureFieldCounter` as well, where it should be `this.loadedFile[severity]['type'][colName]` instead.

**Comment by philboeselager** [comment]
> I think it would be best to just replace error with errors completely and have something like `<div v-for="er in errors" class="ma-2" > {{ er }} </div>` instead.

Yes. I'm going to check this out.

> Also in my test I noticed that a KBART file I used had another tab at the end of each line (except the first one) and therefore all rows threw an error (with 27 instead of 26 columns counted). Not sure if we want to ignore this and instead only look at fewer tabs.

I would recommend NOT to ignore additional tabs, because in some cases it might be difficult to determine, where exactly in the line the additional tab has been entered. I have already seen cases, where the additional tab had been somewhere in the middle of a line and hence had pushed all following content to its adjacent right-sided column.

> I found one other issue not connected to this in ensureFieldCounter as well, where it should be this.loadedFile[severity]['type'][colName] instead.

I'm going to fix it, @hornmo .

**Comment by philboeselager** [comment]
Looks fine to me. Re-converting to pull request.

File: src/shared/popups/gokb-kbart-import-popup/gokb-kbart-import-popup.vue [hunk 1]
```diff
@@ -328,29 +331,51 @@
...
+        }
+        const columnsCount = csvDataRows[0].split(/\t/).length
+        const wrongColumnSizes = new Map()
+        csvDataRows.forEach(function (row, i) {
+          var rowLength = row.split(/\t/).length
+          if (rowLength != columnsCount)
+          wrongColumnSizes.set(i, rowLength)
+        });
+        if (wrongColumnSizes.size != 0) {
+          console.log('kbart.errors.containsDifferentLengthedRows :', this.$i18n.t('kbart.errors.containsDifferentLengthedRows'))
```
Comment (hornmo) [inline src/shared/popups/gokb-kbart-import-popup/gokb-kbart-import-popup.vue]: Don't think this log is necessary any more.

File: src/shared/popups/gokb-kbart-import-popup/gokb-kbart-import-popup.vue [hunk 2]
```diff
@@ -328,29 +331,51 @@
...
+          this.errors.push(this.$i18n.t('kbart.errors.containsDifferentLengthedRows'))
```
Comment (hornmo) [inline src/shared/popups/gokb-kbart-import-popup/gokb-kbart-import-popup.vue]: This message has to be entered into the message JSON files (in `src/locales`). Something more concise like `inconsistentTabsCount` or even just `tabsCount` would be preferable I think. Important is the description (and maybe an instruction how to fix this).

File: src/shared/popups/gokb-kbart-import-popup/gokb-kbart-import-popup.vue [hunk 3]
```diff
@@ -496,21 +520,20 @@
...
+          this.loadedFile[severity][type] = {}
         }
-
-        if (!this.loadedFile[severity][colName][type]) {
-          this.loadedFile[severity][colName][type] = 1
+        if (!this.loadedFile[severity][type][colName]) {
+          this.loadedFile[severity][type][colName] = 1
         } else {
-          this.loadedFile[severity][colName][type]++
+          this.loadedFile[severity][type][colName]++
```
Comment (hornmo) [inline src/shared/popups/gokb-kbart-import-popup/gokb-kbart-import-popup.vue]: This does not conform to the layout in data: 

```
        loadedFile: {
          columns: [],
          errors: {
            missingColumns: [],
            single: [],
            type: {}
          },
          warnings: {
            single: [],
            type: {}
          },
          lineStats: {
            total: 0,
            warning: 0,
            error: 0
          }
        },
```
We want to create a new key for the object `this.loadedFile.error.type`, for example `this.loadedFile.error.type.publication_type.invalid = 1`. So we want to group these first by column and only then by the type of error.

**Review by hornmo — CHANGES_REQUESTED** [review]

File: src/shared/popups/gokb-kbart-import-popup/gokb-kbart-import-popup.vue (hunk 2 above)
Comment (philboeselager) [reply src/shared/popups/gokb-kbart-import-popup/gokb-kbart-import-popup.vue]: Forgot to commit the messages. Changing to `tabsCount`. A fixing instruction is quite complicated and seems too long for this place. Do we use hoverings for these informations?

File: src/shared/popups/gokb-kbart-import-popup/gokb-kbart-import-popup.vue (hunk 3 above)
Comment (philboeselager) [reply src/shared/popups/gokb-kbart-import-popup/gokb-kbart-import-popup.vue]: You seem to have a precise idea whereas I don't understand the point. It's easier if you change this, @hornmo . I'm interested in your changes. You can also remove my commit - these changes have been done separately.

Comment (philboeselager) [reply src/shared/popups/gokb-kbart-import-popup/gokb-kbart-import-popup.vue]: Thanks, I've seen [2b4681a](https://github.com/openlibraryenvironment/gokb-ui/pull/71/commits/2b4681a1e6fd43baaa4790c2d40b04b22f7f77e9). If this is it, fine.

File: src/shared/popups/gokb-kbart-import-popup/gokb-kbart-import-popup.vue (hunk 2 above)
Comment (philboeselager) [reply src/shared/popups/gokb-kbart-import-popup/gokb-kbart-import-popup.vue]: I'm fine with [70e224b](https://github.com/openlibraryenvironment/gokb-ui/pull/71/commits/70e224bd5530da637d519d2b74b3d853d02abf41)

File: src/shared/popups/gokb-kbart-import-popup/gokb-kbart-import-popup.vue (hunk 1 above)
Comment (philboeselager) [reply src/shared/popups/gokb-kbart-import-popup/gokb-kbart-import-popup.vue]: Removed.

**Review by hornmo — APPROVED** [review]

**Pull request merged**